### DISCUSSION TITLE
Fastapi download video with StreamingResponse cause high cpu usage

### DISCUSSION THREAD

**Discussion by pingchesu** [fastapi/fastapi#6050] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import os
import uvicorn
from multiprocessing import current_process
from starlette.background import BackgroundTask
from fastapi.middleware.cors import CORSMiddleware
from fastapi import FastAPI
from fastapi import (APIRouter, BackgroundTasks, Depends, File, HTTPException,
                     Query, Request, UploadFile, status)
from fastapi.responses import (FileResponse, RedirectResponse, Response,
                               StreamingResponse)




def send_bytes_range_requests(
    file_obj, start: int, end: int, chunk_size: int = 10_000
    ):
    """Send a file in chunks using Range Requests specification RFC7233

    `start` and `end` parameters are inclusive due to specification
    """
    with file_obj as f:
        f.seek(start)
        pos = f.tell()
        while pos <= end:
            read_size = min(chunk_size, end + 1 - pos)
            yield f.read(read_size)

def get_range_header(range_header: str, file_size: int):
    try:
        h = range_header.replace("bytes=", "").split("-")
        start = int(h[0]) if h[0] != "" else 0
        end = int(h[1]) if h[1] != "" else file_size - 1
    except ValueError:
        raise HTTPException(
        status.HTTP_416_REQUESTED_RANGE_NOT_SATISFIABLE,
        detail=f"Invalid request range (Range:{range_header})",
    )

    if start > end or start < 0 or end > file_size - 1:
        raise HTTPException(
        status.HTTP_416_REQUESTED_RANGE_NOT_SATISFIABLE,
        detail=f"Invalid request range (Range:{range_header})",
    )
    return start, end

def range_requests_response(
    request: Request, file_path: str, content_type: str, file_name: str = None,
):
    """Returns StreamingResponse using Range Requests of a given file"""
    file_buffer = open(file_path, mode="rb")
#     if file_path.startswith(StorageType.HTTP):
#         file_size = file_buffer.content_length
#     else:
    file_size = os.stat(file_path).st_size
    range_header = request.headers.get("range")

    headers = {
        "content-type": content_type,
        "accept-ranges": "bytes",
        "content-encoding": "identity",
        "content-length": str(file_size),
        "access-control-expose-headers": (
            "content-type, accept-ranges, content-length, "
            "content-range, content-encoding"
        ),
    }
    if file_name:
#         file_name = file_name.encode('utf-8')
        headers['Content-Disposition'] = f"attachment; filename={file_name}"
    start = 0
    end = file_size - 1
    status_code = status.HTTP_200_OK

    if range_header is not None:
        start, end = get_range_header(range_header, file_size)
        size = end - start + 1
        headers["content-length"] = str(size)
        headers["content-range"] = f"bytes {start}-{end}/{file_size}"
        status_code = status.HTTP_206_PARTIAL_CONTENT
    return StreamingResponse(
        send_bytes_range_requests(file_buffer, start, end, 1000*1024),
        headers=headers,
        media_type=content_type,
        background=BackgroundTask(file_buffer.close),
        status_code=status_code,
    )

def create_media_file_router():
    router = APIRouter()
    
    @router.get("/test/media")
    def get_media(request: Request):
        file_path = "test.mp4"
        content_type = 'application/octet-stream'
        file_name = "my_download_filename.mp4"
        return range_requests_response(request, file_path=file_path, content_type='application/octet-stream', file_name=file_name)
    return router

app = FastAPI(version="0.0")
router = create_media_file_router()
app.include_router(router, prefix="/api", tags=["File media"])
origins = ["*"]
app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
    )
if __name__ == '__main__':
    print(f"start fastapi at pid: {os.getpid()}")
    uvicorn.run("main:app",
                host="0.0.0.0",
                port=5000,
                workers=2,
                log_config=None,
                access_log=True,
                forwarded_allow_ips='*',
                proxy_headers=True,
               )
    
# docker run -itd --name fastapi-test -p 5000:5000 -v $(pwd):/app -w /app python:3.8.13-slim-buster /bin/bash -c "pip3 install fastapi==0.83.0 uvicorn==0.20.0 && python3 main_test.py"
```

### Description

HI when I downloaded video with fastapi StreamingResponse it will cause really high cpu usage after video be donwloaded completed! I already checked this [issue](https://github.com/tiangolo/fastapi/issues/1342) but can not find any solution. However, when I download video behind the nginx, the problem disappered. Can someone help me fix this problem?

Reproduce step:

1. pip3 install fastapi==0.83.0 uvicorn==0.20.0
2. put one mp4 file in your current working directory and name it as test.mp4
3. name example code as main.py and run python3 main.py
4. navigate to browser: http://localhost:5000/api/test/media
5. after downloading the video you can use top command to check the cpu usage, it will extremly high

docker reporduce step:
1. download this [link](https://drive.google.com/file/d/1rjS9l0E4Ji51lsEoEcVCB33qDOXjZhNA/view?usp=share_link)
2. unzip fastapi.zip && cd test-fastapi
3. pip3 install docker-compose -> make sure you can use docker-compose command
4. docker-compose up -d -> launch fastapi container service
5. navigate to http://localhost:5002/api/test/media -> this path will directly call fastapi
6. docker stats fastapi-test -> after video download, the cpu usage will increase to 100%
7. docker restart fastapi-test -> reset the container cpu usage
8. navigate to http://localhost:5000/api/test/media -> this path will call fastapi behind nginx
9. docker stats fastapi-test -> after video download, the cpu usage will not increase anymore

Can someone help me fix this problem? 

### Operating System

Linux

### Operating System Details

Same result on running with docker container

### FastAPI Version

0.83.0

### Python Version

3.8

**Comment by iudeen** [comment]
Can you elaborate on "when I download using Nginx, problem disappeared"? Are you directly downloading via nginx or nginx is setup as reverse proxy to FastAPI app?

**Comment by pingchesu** [comment]
> Can you elaborate on "when I download using Nginx, problem disappeared"? Are you directly downloading via nginx or nginx is setup as reverse proxy to FastAPI app?

@iudeen Sorry that I did not make a clear elaboration. When nginx is setup as reverse proxy to FastAPI app, the problem disappeared. You can check my docker reproduce step. I setup nginx as reverse proxy for fastapi app. Thank!

**Comment by Lenclove** [comment]
I tried your code without `nginx`. I think your code is in an endless loop with a function which is called `send_bytes_range_requests`.

**Comment by pythonweb2** [comment]
Nginx would be the preferred method to stream content like videos, using features like X-Accel-Redirect (sometimes known as X-SendFile) if authentication is needed. Is there any reason why you would want to stream the video using FastAPI?

**Comment by panla** [comment]
use Nginx support an url of video for frontend, donwload it by frontend, not backend api.
